### PR TITLE
nixos/keycloak: Escape initial admin password and database password properly

### PR DESCRIPTION
**Pull request by talyz**
###### Description of changes
Using special characters such as `'`, `"`, `>` or `;` in either the initial admin password, database password or both currently lead to failures, since they're not properly escaped and therefore interpreted by the shell or the database. Fix this by using `lib.escapeShellArg` when splicing in the admin password and doubling any single quotes in the database password.

The issue was raised in #195142. This provides an alternative solution.

cc @red-hood

Closes: #195142

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/keycloak.nix [hunk 1]
```diff
@@ -544,7 +544,12 @@ in
             create_role="$(mktemp)"
             trap 'rm -f "$create_role"' EXIT
 
+            # Read the password from the credentials directory and
+            # escape any single quotes by adding additional single
+            # quotes after them.
             db_password="$(<"$CREDENTIALS_DIRECTORY/db_password")"
+            db_password="''${db_password//\'/\'\'}"
+
```
Comment (infinisil) [inline nixos/modules/services/web-apps/keycloak.nix:552]: Why is this necessary?

Comment (talyz) [reply nixos/modules/services/web-apps/keycloak.nix:552]: It's to escape any single quotes in the password which would otherwise break the quoting of the password in the SQL snippet further down. Single quotes can be escaped in SQL by adding an additional single quote in front of them, so that's what this does.

Comment (infinisil) [reply nixos/modules/services/web-apps/keycloak.nix:552]: Ahh I see, didn't read that far, makes sense! Ideally we wouldn't have to do this escaping ourselves, but from some research it looks like there's no other way for postgresql. I think we should at least link to [postgresql's string escaping rules](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-CONSTANTS) though. Also check out that caution ssign there, it seems to imply that the string escaping can change depending on the [standard_conforming_strings](https://www.postgresql.org/docs/current/runtime-config-compatible.html#GUC-STANDARD-CONFORMING-STRINGS) parameter, whose default changed in the past.. This could mess the escaping here up :neutral_face:

Comment (red-hood) [reply nixos/modules/services/web-apps/keycloak.nix:552]: Thank you for the implementation, I did not have time so far to continue with the issue

Comment (talyz) [reply nixos/modules/services/web-apps/keycloak.nix]: Okay, I've added links in the comments to the relevant pages and also an assertion that `standard_conforming_strings` are not disabled.

I also found out that MySQL allows escape sequences by default and that has to be turned off temporarily for the expected behavior. And, when testing for that, I found that escape sequences are also interpreted somewhere in Keycloak or, more likely, the JDBC driver and have to be escaped in the config file.. *sigh*

@red-hood No problem! Thank you for raising the issue! :)

**Review by mweinelt — APPROVED** [review]
